Additional "default browser" prompts: variant 2

LukasPaczos opened the pull request:
Task/Issue URL: https://app.asana.com/0/1208671518894266/1208944504536347/f

### Description
Introduces variant 2 of the experiment that proposes new CTAs around setting DDG as a default browser. The PR also adds all of the experiment foundations and abstractions, so the remaining two variants will only be a small formality.

I'm going to follow up on pixels in a separate PR.

### Class diagram
![image](https://github.com/user-attachments/assets/8a8f5c84-0d73-4024-ac72-bd4650556ab0)

### Flow diagram
![image](https://github.com/user-attachments/assets/1846fe61-8acb-4a63-a297-cd4d9c907145)

### Steps to test this PR

This PR relies on the new A/B/N framework for rollout, so to test you'll need a custom privacy config which includes the feature definition, for example:
```json
"defaultBrowserPrompts": {
  "state": "enabled",
  "exceptions": [],
  "features": {
    "additionalPrompts": {
      "state": "enabled",
      "rollout": {
        "steps": [
          {
            "percent": 100
          }
        ]
      },
      "settings": {
        "activeDaysUntilStage1": 1,
        "activeDaysUntilStage2": 3,
        "activeDaysUntilStop": 5
      },
      "cohorts": [
        {
          "name": "control",
          "weight": 0
        },
        {
          "name": "variant_2",
          "weight": 1
        }
      ]
    }
  },
}
```

or you can apply this JSON Blob:
```diff
diff --git a/privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyFeatureName.kt b/privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyFeatureName.kt
index ca0870311..5a904f76b 100644
--- a/privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyFeatureName.kt
+++ b/privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyFeatureName.kt
@@ -27,4 +27,5 @@ enum class PrivacyFeatureName(val value: String) {
     TrackingParametersFeatureName("trackingParameters"),
 }
 
-const val PRIVACY_REMOTE_CONFIG_URL = "https://staticcdn.duckduckgo.com/trackerblocking/config/v4/android-config.json"
+// const val PRIVACY_REMOTE_CONFIG_URL = "https://staticcdn.duckduckgo.com/trackerblocking/config/v4/android-config.json"
+const val PRIVACY_REMOTE_CONFIG_URL = "https://www.jsonblob.com/api/1329040942036082688"
```

Once built with the right config, ensure that you **don't have DDG app set as a default browser**, only then you'll be assigned to the experiment.

Given the example config above:
- after one active day you'll see a dialog pop up
- after additional 2 active days you'll see a highlighted menu icon and a new menu button
- after additional 2 days, the menu button will disappear

If you convert at any point and set DDG app as default browser, the experiment will permanently stop.

[Screen_recording_20250115_173728.webm](https://github.com/user-attachments/assets/a63f243d-b21f-4dc5-9e69-53e6ee90524b)

In the video I'm using the [ADB commands to change the date](https://app.asana.com/0/1208671518894266/1209150522527594/f), but any mechanism of date change will be enough :)

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    override fun onMessageDialogConfirmationButtonClicked() {
+        appCoroutineScope.launch {
```
dispatchers?

LukasPaczos replied:
I actually was able to remove this coroutine dispatch altogether.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+
+    override fun onSystemDefaultBrowserDialogShown() {
+        browserSelectionWindowFallbackDeferred = appCoroutineScope.async {
+            delay(FALLBACK_TO_DEFAULT_APPS_SCREEN_THRESHOLD_MILLIS)
```
some explanation for this delay would be helpful.

LukasPaczos replied:
There's an explanation attached to the `browserSelectionWindowFallbackDeferred` field at the top of the class, let me know if you think we need more than that.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+                featureSettingsJsonAdapter.fromJson(settings)
+            } catch (e: Exception) {
+                Timber.e(e)
+                null
```
I’m not sure about returning null when there’s an issue parsing. How about providing a set of default config?

LukasPaczos replied:
Same as https://github.com/duckduckgo/Android/pull/5476#discussion_r1933600731. The experiment is fully controlled via config, so if the config doesn't work, I'd rather escape than hardcode defaults.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/diagrams/default_browser_prompts_experiment_class_diagram.puml:
I’m not opposed to this, but let’s discuss in the team meeting what to do about it. code should be self explanatory (it pretty much is in this case). We tend to add documentation in Asana and not here, so I’d like to agree on the use / need for UML files.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/omnibar/OmnibarLayoutViewModel.kt:
```diff
@@ -126,6 +130,16 @@ class OmnibarLayoutViewModel @Inject constructor(
         GLOBE,
     }
 
+    init {
```
I’d move this to `onAttachedToWindow`. There are cases where you’ll lose this ViewState update just because the view is not yet attached.

malmstein reviewed (approved):
This looks great, nicely done @LukasPaczos. I’ve left a few comments, most of them are nits besides the dispatchers.

LukasPaczos commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+                parsedSettings
+            } ?: run {
+                Timber.e("Failed to obtain feature settings.")
+                return
```
If we ever land in this code branch, there's something very wrong going on with the config, so I'd rather escape from the experiment than hardcode stage intervals.

LukasPaczos commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -1046,6 +1046,7 @@ class BrowserTabFragment :
         binding.rootView.postDelayed(POPUP_MENU_DELAY) {
             if (isAdded) {
                 popupMenu.show(binding.rootView, omnibar.toolbar)
+                viewModel.onPopupMenuLaunched()
```
I see this as a generic view model callback, so I think it can react to both Custom Tabs and regular Browser Activity. In case of the experiment that is later invoked form the view model, it's fine to get called even if the browser is already default.

LukasPaczos commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+        initialValue = false,
+    )
+
+    @VisibleForTesting
```
I intended for this to be visible for testing but not being accessible to any other class. It is a tiny smell, and I could wrap all Moshi interactions in a separate class so that I can mock it, but not sure if it's worth it. Let me know if you have a strong opinion.

LukasPaczos commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+        }
+    }
+
+    override fun onMessageDialogShown() {
```
yup, I'll clean up whatever ends up not being needed in the pixels PR

LukasPaczos commented on app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    override fun onOverflowMenuOpened() {
+        appCoroutineScope.launch {
```
I updated the data store to switch context to IO.

LukasPaczos commented:
Thanks for the comments @malmstein! I renamed all "overflow menu" references to "popup menu" and removed/updated some coroutine dispatchers. I want the evaluations to run sequentially and I'm happy to use the `default` dispatcher but I added a mutex for the `evaluate` function.

LukasPaczos commented:
We synced internally that we can start merging and will retest manually again before enabling the feature flags.
